drivers/led/led_dac.c

```diff
@@ -0,0 +1,132 @@
+
+	if (led >= config->num_leds) {
+		return -EINVAL;
+	}
+
+	return led_dac_set_raw(dev, led, 0);
+}
+
+static DEVICE_API(led, led_dac_api) = {
+	.on = led_dac_on,
+	.off = led_dac_off,
+	.set_brightness = led_dac_set_brightness,
+};
+
+static int led_dac_init(const struct device *dev)
+{
+	const struct led_dac_config *config = dev->config;
+	int ret;
+
+	for (uint8_t i = 0; i < config->num_leds; ++i) {
+		const struct led_dac_leds *led = &config->leds[i];
+
+		if (!device_is_ready(led->dac)) {
+			return -ENODEV;
+		}
+
+		ret = dac_channel_setup(led->dac, &led->chan_cfg);
+		if (ret != 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+#define LED_DAC_MAX(n) (BIT(DT_PROP(n, resolution)) - 1)
+
+#define LED_DAC_DT_GET(n)                                                                          \
+	{                                                                                          \
+		.dac = DEVICE_DT_GET(DT_PHANDLE(n, dac)),                                          \
+		.chan_cfg = {.channel_id = DT_PROP(n, channel),                                    \
+			     .resolution = DT_PROP(n, resolution),                                 \
+			     .buffered = false,                                                    \
+			     .internal = false},                                                   \
+		.dac_max = LED_DAC_MAX(n), .dac_min = LED_DAC_MIN(n)                               \
+	}
+
+#define LED_DAC_DEFINE(n, compat)                                                                  \
+	static const struct led_dac_leds led_##compat##_##n[] = {                                  \
+		DT_INST_FOREACH_CHILD_SEP(n, LED_DAC_DT_GET, (,))};                                \
+                                                                                                   \
+	static const struct led_dac_config led_##compat##_config_##n = {                           \
+		.leds = led_##compat##_##n,                                                        \
+		.num_leds = ARRAY_SIZE(led_##compat##_##n),                                        \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(n, &led_dac_init, NULL, NULL, &led_##compat##_config_##n,            \
+			      POST_KERNEL, CONFIG_LED_INIT_PRIORITY, &led_dac_api);
+
+#define DT_DRV_COMPAT dac_leds
+#define LED_DAC_MIN(n)                                                                             \
+	((uint64_t)LED_DAC_MAX(n) *                                                                \
+	 ((uint64_t)DT_PROP(n, forward_voltage_mv) + DT_PROP(n, resistor_forward_voltage_mv)) /    \
+	 DT_PROP(n, max_voltage_mv))
+DT_INST_FOREACH_STATUS_OKAY_VARGS(LED_DAC_DEFINE, DT_DRV_COMPAT);
+
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT dac_leds_vccs
+#undef LED_DAC_MIN
+#define LED_DAC_MIN(n) 0
```

Comment:
Honestly, I don't like the idea of having two distinct bindings for LED DAC devices. The code above is confusing and quite messy.

Here we are just computing `min` and `max` values. What about letting the user do that with some dedicated DT properties ? With maybe an help message in the binding file explaining how to compute them ?

Reply:
It looks good now.